### ISSUE TITLE
Relative URL isn't appearing in mastodon

### ISSUE THREAD

**Issue by aciccarello** [snarfed/bridgy-fed#390]
I posted [a note with a relative URL](https://www.ciccarello.me/drafts) linking to another page on my site.

```html
<div class="post-content mdc-typography--body1 e-content">
  <p>
    I added a new page to my site this week. You can now preview some of the posts I have 
    in progress on the <a href="/drafts">/drafts</a> page!
  </p>
  <p>Beware, I don’t plan on redirecting those old URLs though so I don’t recommend linking to draft posts.</p>
</div>
```

However, when this appeared in Mastodon the link was removed. ([url for logged in user](https://toot.cafe/@anthony@ciccarello.me/109764425770951643))
<img width="593" alt="screenshot of mastodon web client showing note text with no link." src="https://user-images.githubusercontent.com/11273838/215242521-6c5efbd6-7a8c-41a3-a167-3e94f6f96c0d.png">

Could bridgy fed convert the relative path to absolute before sending?

**Comment by snarfed** [comment]
Yes! Ideally it should. We'll see how easy it is in practice...

**Comment by snarfed** [comment]
This isn't looking easy, at least with the tools I have. mf2py fully qualifies URLs it hoists out of content, eg in mf2 properties and rel URLs, but not inside `<a>` tags in HTML `e-content`. BeautifulSoup doesn't either. So I'd need to do it myself, or find a new tool to do it. Ugh.

Interestingly, php-mf2 _does_ fully qualify link URLs inside HTML content, eg "avoid it altogether" in https://pin13.net/mf2/?url=https://snarfed.org/2023-01-11_logs-as-end-user-ui . Hmm! That should would be a nice mf2py feature. cc @sknebel @angelogladding @tommorris

**Comment by barnabywalters** [comment]
IIRC we implemented this in php-mf2 because it was easy, useful and IMO implied by the mf2 parsing algorithm requiring any other relative URLs to be resolved. Embedded content is presumably going to be reused out of context, and it’s not particularly useful if it contains URLs which were relative to some other context! I agree with Zegnat’s proposal that it get explicitly added to the parsing spec. I’m surprised it hasn’t come up in tests already, to be honest.

**Comment by aciccarello** [comment]
I haven't tested it yet but I think this could also affect #45 in that hashtag URL need to be absolute.

**Comment by snarfed** [comment]
Sounds like we might get this in mf2py soon! https://github.com/microformats/mf2py/issues/197#issuecomment-1811675685

**Comment by snarfed** [comment]
This shipped just now in v2.0.1!

**Comment by aciccarello** [comment]
Just re-sent the mention for the original post and I can now see the URL correctly in Mastodon. 👏
